parsl/dataflow/dflow.py

```diff
@@ -920,10 +921,14 @@ def wait_for_current_tasks(self):
             if task_id not in self.tasks:
                 logger.debug("Task {} no longer in task list".format(task_id))
             else:
-                fut = self.tasks[task_id]['app_fu']
+                task_record = self.tasks[task_id]  # still a race condition with the above self.tasks if-statement
+                fut = task_record['app_fu']
                 if not fut.done():
-                    logger.debug("Waiting for task {} to complete".format(task_id))
                     fut.exception()
+                # now app future is done, poll until DFK state is final: a DFK state being final and the app future being done do not imply each other.
+                while task_record['status'] not in FINAL_STATES:
```

Comment:
I believe it is unlikely that we'll be stuck in an infinite loop here, have you seen any hangs over here?

Reply:
I haven't seen hangs. I think one would happen if the post completion code broke before setting the task record status.